beacon-chain/verification/data_column.go

```diff
@@ -0,0 +1,506 @@
+			return columnErrBuilder(ErrSidecarNotFinalizedDescendent)
+		}
+	}
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarInclusionProven() (err error) {
+	if ok, err := dv.results.cached(RequireSidecarInclusionProven); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarInclusionProven, &err)
+
+	startTime := time.Now()
+
+	for _, dataColumn := range dv.dataColumns {
+		k, keyErr := inclusionProofKey(dataColumn)
+		if keyErr == nil {
+			if _, ok := dv.ic.Get(k); ok {
+				continue
+			}
+		} else {
+			log.WithError(keyErr).Error("Failed to get inclusion proof key")
+		}
+
+		if err = peerdas.VerifyDataColumnSidecarInclusionProof(dataColumn); err != nil {
+			return columnErrBuilder(ErrSidecarInclusionProofInvalid)
+		}
+
+		if keyErr == nil {
+			dv.ic.Add(k, struct{}{})
+		}
+	}
+
+	dataColumnSidecarInclusionProofVerificationHistogram.Observe(float64(time.Since(startTime).Milliseconds()))
+
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarKzgProofVerified() (err error) {
+	if ok, err := dv.results.cached(RequireSidecarKzgProofVerified); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarKzgProofVerified, &err)
+
+	startTime := time.Now()
+
+	err = dv.verifyDataColumnsCommitment(dv.dataColumns)
+	if err != nil {
+		return columnErrBuilder(errors.Wrap(err, "verify data column commitment"))
+	}
+
+	dataColumnBatchKZGVerificationHistogram.Observe(float64(time.Since(startTime).Milliseconds()))
+	return nil
+}
+
+func (dv *RODataColumnsVerifier) SidecarProposerExpected(ctx context.Context) (err error) {
+	if ok, err := dv.results.cached(RequireSidecarProposerExpected); ok {
+		return err
+	}
+
+	defer dv.recordResult(RequireSidecarProposerExpected, &err)
+
+	for _, dataColumn := range dv.dataColumns {
+		// Extract the slot of the data column.
+		dataColumnSlot := dataColumn.Slot()
+
+		// Compute the epoch of the data column slot.
+		dataColumnEpoch := slots.ToEpoch(dataColumnSlot)
+		if dataColumnEpoch > 0 {
+			dataColumnEpoch = dataColumnEpoch - 1
+		}
+
+		// Extract the root of the parent block corresponding to the data column.
+		parentRoot := dataColumn.ParentRoot()
+
+		// Compute the target root for the epoch.
+		targetRoot, err := dv.fc.TargetRootForEpoch(parentRoot, dataColumnEpoch)
+		if err != nil {
+			return columnErrBuilder(ErrSidecarUnexpectedProposer)
+		}
+
+		// Create a checkpoint for the target root.
+		checkpoint := &forkchoicetypes.Checkpoint{Root: targetRoot, Epoch: dataColumnEpoch}
+
+		// Try to extract the proposer index from the data column in the cache.
+		idx, cached := dv.pc.Proposer(checkpoint, dataColumnSlot)
+
+		if !cached {
+			// Retrieve the root of the parent block corresponding to the data column.
+			parentRoot := dataColumn.ParentRoot()
+
+			// Retrieve the parentState state to fallback to full verification.
+			parentState, err := dv.sr.StateByRoot(ctx, parentRoot)
+			if err != nil {
+				return columnErrBuilder(ErrSidecarUnexpectedProposer)
+			}
```

Comment:
Same question as before, the amount of copies we do here may be excessive

Reply:
Fixed in 58c70f27802a95cd385496ff0c67eaef06f11888 and 74f1300fb07d7c53c539f1b7219b5cb34b5b1b48.